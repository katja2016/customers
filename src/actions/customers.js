import database from '../firebase/firebase';

// ADD_CUSTOMER
export const addCustomer = (customer) => {
    return {
        type: 'ADD_CUSTOMER',
        customer
    };
};

export const startAddCustomer = (customerData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = customerData;
        const customer = { description, note, amount, createdAt };

        return database.ref(`users/${uid}/customers`).push(customer).then((ref) => {
            dispatch(addCustomer({
                id: ref.key,
                ...customer
            }));
        });
    };
};

// REMOVE_CUSTOMER
export const removeCustomer = ({ id } = {}) => {
    return {
        type: 'REMOVE_CUSTOMER',
        id
    };
};

export const startRemoveCustomer = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/customers/${id}`).remove().then(() => {
           dispatch(removeCustomer({ id }));
        });
    };
};

// EDIT_CUSTOMER
export const editCustomer = (id, updates) => {
    return {
        type: 'EDIT_CUSTOMER',
        id,
        updates
    };
};

export const startEditCustomer = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/customers/${id}`).update(updates).then(() => {
            dispatch(editCustomer(id, updates));
        });
    };
};

// SET_CUSTOMERS
export const setCustomers = (customers) => ({
    type: 'SET_CUSTOMERS',
    customers
});

export const startSetCustomers = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/customers`).once('value').then((snapshot) => {
            const customers = [];
            snapshot.forEach((childSnapshot) => {
                customers.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });

            dispatch(setCustomers(customers));
        });
    };
};