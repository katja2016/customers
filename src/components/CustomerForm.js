import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class CustomerForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.customer ? props.customer.description : '',
            note: props.customer ? props.customer.note : '',
            amount: props.customer ? (props.customer.amount).toString() : '',
            createdAt: props.customer ? moment(props.customer.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = (e) => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/g)){
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        if(createdAt){
            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            this.setState(() => ({ error: 'Enter valid data for description and amount fields.' }));
        }else{
            this.setState(() => ({ error: '' }));
            
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && <div className="form__error">{this.state.error}</div>}
                <input
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    className="text-input"
                />
                <input
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    className="text-input"
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    placeholder="Add a note about your customer"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                    className="textarea"
                ></textarea>
                <div>
                    <button className="button">Save</button>
                </div>
            </form>
        );
    };
}