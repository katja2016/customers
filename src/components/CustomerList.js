import React from 'react';
import { connect } from 'react-redux';
import CustomerListItem from './CustomerListItem';
import selectCustomers from '../selectors/customers';

export const CustomerList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Customers</div>
            <div className="show-for-desktop">Customer</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {props.customers.length === 0 ? (
                <div>
                    <span className="list-item list-item--message">No customers</span>
                </div>
            ) : (
                props.customers.map((customer) => {
                    return <CustomerListItem {...customer} key={customer.id}/>;
                }))}
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        customers: selectCustomers(state.customers, state.filters),
    };
};

export default connect(mapStateToProps)(CustomerList);
