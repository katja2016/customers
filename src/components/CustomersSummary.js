import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectCustomers from '../selectors/customers';
import selectCustomersTotal from '../selectors/customers-total';
import numeral from 'numeral';

export const CustomersSummary = ({ customersCount, customersTotal }) => {
    const customerWord = customersCount === 1 ? 'customer' : 'customers';
    const formattedCustomersTotal = numeral(customersTotal).format('$0,0.00');

    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span>{customersCount}</span> {customerWord} spending <span>{formattedCustomersTotal}</span></h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add customer</Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleCustomers = selectCustomers(state.customers, state.filters);
    return {
        customersCount: visibleCustomers.length,
        customersTotal: selectCustomersTotal(visibleCustomers)
    };
};

// connect to redux store
export default connect(mapStateToProps)(CustomersSummary);
