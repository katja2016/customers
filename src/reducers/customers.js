const customersReducer = (state = [], action) => {
    switch (action.type){
        case 'ADD_CUSTOMER':
            return [...state, action.customer];
        case 'REMOVE_CUSTOMER':
            return state.filter((e) => e.id !== action.id);
        case 'EDIT_CUSTOMER':
            return state.map((customer) => {
                if(customer.id === action.id){
                    return { ...customer, ...action.updates };
                }else{
                    return customer;
                }
            });
        case 'SET_CUSTOMERS':
            return action.customers;
        default:
            return state;
    }
};

export default customersReducer;