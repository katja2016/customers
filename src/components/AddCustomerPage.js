import React from 'react';
import { connect } from 'react-redux';
import CustomerForm from './CustomerForm';
import { startAddCustomer } from '../actions/customers';


export class AddCustomerPage extends React.Component {
    onSubmit = (customer) => {
        this.props.startAddCustomer(customer);
        this.props.history.push('/');
    };

    render() {
        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Customer</h1>
                    </div>
                </div>
                <div className="content-container">
                    <CustomerForm onSubmit={this.onSubmit} />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startAddCustomer: (customer) => dispatch(startAddCustomer(customer))
    };
};

export default  connect(undefined, mapDispatchToProps)(AddCustomerPage);
