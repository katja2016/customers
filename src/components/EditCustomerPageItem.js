import React from 'react';
import { connect } from 'react-redux';
import CustomerForm from './CustomerForm';
import { startEditCustomer, startRemoveCustomer } from '../actions/customers';

export class EditCustomerPageItem extends React.Component {

    onSubmit = (customer) => {
        this.props.startEditCustomer(this.props.customer.id, customer);
        this.props.history.push('/');
    };

    onRemove = () => {
        this.props.startRemoveCustomer({ id: this.props.customer.id });
        this.props.history.push('/');
    };

    render(){
        return(
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Customer</h1>
                    </div>
                </div>
                <div className="content-container">
                    <CustomerForm
                        customer={this.props.customer}
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.onRemove}>Remove Customer</button>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        customer: state.customers.find((espense) => {
            return espense.id === props.match.params.id;
        })
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        startEditCustomer: (id, customer) => dispatch(startEditCustomer(id, customer)),
        startRemoveCustomer: (data) => dispatch(startRemoveCustomer(data))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditCustomerPageItem);