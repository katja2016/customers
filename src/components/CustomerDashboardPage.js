import React from 'react';
import CustomerList from './CustomerList';
import CustomerListFilters from './CustomerListFilters';
import CustomersSummary from './CustomersSummary';

const CustomerDashboardPage = () => (
    <div>
        <CustomersSummary/>
        <CustomerListFilters />
        <CustomerList />
    </div>
);

export default CustomerDashboardPage;
