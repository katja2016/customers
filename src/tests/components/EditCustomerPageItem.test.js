import React from 'react';
import { shallow } from 'enzyme';
import customers from '../fixtures/customers';
import { EditCustomerPageItem } from '../../components/EditCustomerPageItem';

let startEditCustomer, startRemoveCustomer, history, wrapper;

beforeEach(() => {
    startEditCustomer = jest.fn();
    startRemoveCustomer = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(
        <EditCustomerPageItem
            startEditCustomer={startEditCustomer}
            startRemoveCustomer={startRemoveCustomer}
            history={history}
            customer={customers[2]}
        />
    );
});

test('should render EditCustomerPageItem correctly', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should handle startEditCustomer', () => {
    wrapper.find('CustomerForm').prop('onSubmit')(customers[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startEditCustomer).toHaveBeenLastCalledWith(customers[2].id, customers[2]);
});

test('should handle startRemoveCustomer', () => {
    wrapper.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(startRemoveCustomer).toHaveBeenLastCalledWith({ id: customers[2].id });
});