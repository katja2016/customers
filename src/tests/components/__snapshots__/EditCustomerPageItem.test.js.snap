// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditCustomerPageItem correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Customer
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <CustomerForm
      customer={
        Object {
          "amount": 4500,
          "createdAt": 345600000,
          "description": "Credit Card",
          "id": "3",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Customer
    </button>
  </div>
</div>
`;
